metricbeat/module/elasticsearch/index/data_xpack.go

```diff
@@ -0,0 +1,332 @@
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package index
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common"
+	s "github.com/elastic/beats/libbeat/common/schema"
+	c "github.com/elastic/beats/libbeat/common/schema/mapstriface"
+	"github.com/elastic/beats/metricbeat/helper/elastic"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/metricbeat/module/elasticsearch"
+)
+
+var (
+	// Based on https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/collector/indices/IndexStatsMonitoringDoc.java#L127-L203
+	xpackSchema = s.Schema{
+		"uuid":      c.Str("uuid"),
+		"primaries": c.Dict("primaries", indexStatsSchema),
+		"total":     c.Dict("total", indexStatsSchema),
+	}
+
+	indexStatsSchema = s.Schema{
+		"docs": c.Dict("docs", s.Schema{
+			"count": c.Int("count"),
+		}),
+		"fielddata": c.Dict("fielddata", s.Schema{
+			"memory_size_in_bytes": c.Int("memory_size_in_bytes"),
+			"evictions":            c.Int("evictions"),
+		}),
+		"indexing": c.Dict("indexing", s.Schema{
+			"index_total":             c.Int("index_total"),
+			"index_time_in_millis":    c.Int("index_time_in_millis"),
+			"throttle_time_in_millis": c.Int("throttle_time_in_millis"),
+		}),
+		"merges": c.Dict("merges", s.Schema{
+			"total_size_in_bytes": c.Int("total_size_in_bytes"),
+		}),
+		"query_cache":   c.Dict("query_cache", cacheStatsSchema),
+		"request_cache": c.Dict("request_cache", cacheStatsSchema),
+		"search": c.Dict("search", s.Schema{
+			"query_total":          c.Int("query_total"),
+			"query_time_in_millis": c.Int("query_time_in_millis"),
+		}),
+		"segments": c.Dict("segments", s.Schema{
+			"count":                         c.Int("count"),
+			"memory_in_bytes":               c.Int("memory_in_bytes"),
+			"terms_memory_in_bytes":         c.Int("terms_memory_in_bytes"),
+			"stored_fields_memory_in_bytes": c.Int("stored_fields_memory_in_bytes"),
+			"term_vectors_memory_in_bytes":  c.Int("term_vectors_memory_in_bytes"),
+			"norms_memory_in_bytes":         c.Int("norms_memory_in_bytes"),
+			"points_memory_in_bytes":        c.Int("points_memory_in_bytes"),
+			"doc_values_memory_in_bytes":    c.Int("doc_values_memory_in_bytes"),
+			"index_writer_memory_in_bytes":  c.Int("index_writer_memory_in_bytes"),
+			"version_map_memory_in_bytes":   c.Int("version_map_memory_in_bytes"),
+			"fixed_bit_set_memory_in_bytes": c.Int("fixed_bit_set_memory_in_bytes"),
```

Comment:
Not sure about keeping the units in the names :thinking:

Reply:
This is because these events are going into the `.monitoring-es-6-mb-*` indices, which need to have the same structure as existing `.monitoring-es-6-*` indices. Otherwise the UI will need to add special logic depending on whether the data is coming from metricbeat or internal shippers.

If you look at the non-xpack schema for this metricset, we have removed the units from the names (and are also indexing a lot fewer fields): https://github.com/elastic/beats/blob/9828935bb1896ea620a77a328a351fd6e828112f/metricbeat/module/elasticsearch/index/data.go#L33-L51